acme/acme.go

```diff
@@ -328,29 +355,44 @@ func (a *ACME) CreateLocalConfig(tlsConfig *tls.Config, checkOnDemandDomain func
 func (a *ACME) getCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
 	domain := types.CanonicalDomain(clientHello.ServerName)
 	account := a.store.Get().(*Account)
-	//use regex to test for wildcard certs that might have been added into TLSConfig
-	for k := range a.TLSConfig.NameToCertificate {
-		selector := "^" + strings.Replace(k, "*.", "[^\\.]*\\.?", -1) + "$"
-		match, _ := regexp.MatchString(selector, domain)
-		if match {
-			return a.TLSConfig.NameToCertificate[k], nil
-		}
-	}
+
+	// Check for existing challenge cert
+	log.Debugf("Checking ACME challenge for %s", domain)
 	if challengeCert, ok := a.challengeProvider.getCertificate(domain); ok {
-		log.Debugf("ACME got challenge %s", domain)
+		log.Debugf("Returning ACME challenge cert for %s", domain)
 		return challengeCert, nil
 	}
+
+	// Check for existing ACME domain cert
+	log.Debugf("Checking ACME domain cert for %v", domain)
 	if domainCert, ok := account.DomainsCertificate.getCertificateForDomain(domain); ok {
-		log.Debugf("ACME got domain cert %s", domain)
+		log.Debugf("Returning ACME domain cert for %s", domain)
 		return domainCert.tlsCert, nil
 	}
+
 	if a.OnDemand {
 		if a.checkOnDemandDomain != nil && !a.checkOnDemandDomain(domain) {
 			return nil, nil
 		}
-		return a.loadCertificateOnDemand(clientHello)
+		cert, err := a.loadCertificateOnDemand(clientHello)
+		if cert != nil && err == nil {
+			return cert, nil
+		}
+		log.Errorf("On demand certicate retrival failed for %v due to %v", domain, err)
 	}
-	log.Debugf("ACME got nothing %s", domain)
+
+	// Check if we have a wildcard cert into TLSConfig that could be used
+	log.Debugf("Checking wildcard certificate matching for %v", domain)
+	for k := range a.TLSConfig.NameToCertificate {
```

Comment:
Well the issue I tried to address with this PR was also the fact that if I have both a self signed wildcard cert and an ACME cert, only the wildcard will be returned (case 2 and 3 in PR description).

The purpose of the `IgnoreFilters` was to filter out some (sub)domains for LE certificate generation. (case 1 in PR description).

So to sum up I'm trying to address:

1 - if LE and wildcard are present, return LE
2 - the `IgnoreFilters` can be used to filter the LE cert generation

I don't think the part 2 is changing any behavior. Part 1 is.

Now I can see 2 uses cases for Part 1.

- Current behavior, always return wildcard if any
  - if I have a signed wildcard, it makes sense to use it instead of LE cert (but the LE cert will still exists, hidden, and will be renewed and not used).
- New behavior, return LE over wildcard,
  - make sense if I use a self signed wildcard cert.

I'm not sure on how to address those two cases as they can both exists in the same traefik instance.